data-library/library/script/nycoc_checkbook.py

```diff
@@ -0,0 +1,381 @@
+        return parent_node
+    elif isinstance(parent_val, list):
+        for list_item in parent_val:
+            if len(list_item) != 0:
+                parent_node.append(dict_to_xml_obj(list_item))
+        return parent_node
+    else:
+        for child_key, child_val in parent_val.items():
+            if len(child_val) != 0:
+                parent_node.append(dict_to_xml_obj({child_key: child_val}))
+        return parent_node
+
+
+def xml_obj_to_string(xml_obj: ET.Element) -> str:
+    """Returns a XML string representation of python xml object."""
+    return ET.tostring(xml_obj, encoding="utf-8").decode("utf-8")
+
+
+def generate_monthly_ranges(
+    start_period: str, end_period: str, date_format: str = DEFAULT_DATE_FORMAT
+) -> list[tuple[str, str]]:
+    """
+    Generate a list of tuples representing monthly date ranges within the specified time period.
+
+    Example:
+    ```python
+    >>> result = generate_monthly_ranges("2023-07-08", "2023-08-27")
+    >>> print(result)
+    [("2023-07-08", "2023-07-30"), ("2023-08-01", "2023-08-27")]
+    ```python
+    """
+
+    start_period_dt = dt.datetime.strptime(start_period, date_format)
+    end_period_dt = dt.datetime.strptime(end_period, date_format)
+
+    # if start date is later than end date, flip them
+    if start_period_dt > end_period_dt:
+        start_period_dt, end_period_dt = end_period_dt, start_period_dt
+
+    start_month_date = start_period_dt
+    _, num_days = calendar.monthrange(start_month_date.year, start_month_date.month)
+    end_month_date = start_month_date.replace(day=num_days)
+
+    monthly_ranges = []
+
+    while end_month_date <= end_period_dt:
+        monthly_ranges.append(
+            (
+                start_month_date.strftime(date_format),
+                end_month_date.strftime(date_format),
+            )
+        )
+        start_month_date = end_month_date + dt.timedelta(days=1)
+        _, num_days = calendar.monthrange(start_month_date.year, start_month_date.month)
+        end_month_date = start_month_date.replace(day=num_days)
+
+    if end_month_date > end_period_dt:
+        monthly_ranges.append(
+            (
+                start_month_date.strftime(date_format),
+                end_period_dt.strftime(date_format),
+            )
+        )
+
+    return monthly_ranges
+
+
+def create_data_dir(data_path: Path = DEFAULT_DATA_PATH) -> None:
+    """Creates a local data directory."""
+    if data_path.exists():
+        shutil.rmtree(data_path)
+    try:
+        data_path.mkdir(parents=False, exist_ok=False)
+    except Exception as err:
+        print(f"❌ Unable to create data directory.")
+        raise err
+
+
+def read_parquet_files_to_df(data_dir: Path) -> pd.DataFrame:
+    """
+    Reads Parquet files from a given data directory and concatenates
+    them into a single pandas DataFrame.
+    """
+    df = pd.concat(
+        pd.read_parquet(parquet_file) for parquet_file in data_dir.glob("*.parquet")
```

Comment:
Just curious, did you try out pyarrow much? How did that go? This ended up being not quite as "big" data as we initially thought, but still curious on performance and how it would compare to appending as we go with pyarrow.

Again, more curious than anything! This works nicely